src/pybind/mgr/osd_support/module.py

```diff
@@ -0,0 +1,284 @@
+                stderr=f"Command not found <{cmd.get('prefix', '')}>")
+        return HandleCommandResult(
+            retval=ret,   # exit code
+            stdout=out,   # stdout
+            stderr=err)
+
+    def serve(self):
+        """
+        This method is called by the mgr when the module starts and can be
+        used for any background activity.
+        """
+        self.log.info("Starting mgr/osd_support")
+        while self.run:
+            # Do some useful background work here.
+
+            self.log.debug(f"Scheduled for draining: <{self.osd_ids}>")
+            self.log.debug(f"Currently being drained: <{self.emptying_osds}>")
+            # the state should be saved to the mon store in the actual call and
+            # then retrieved in serve() probably
+
+            # 1) check if all provided osds can be stopped, if not, shrink list until ok-to-stop
+            for x in self.find_osd_stop_threshold(self.osd_ids):
+                self.emptying_osds.add(x)
+
+            # remove the emptying osds from the osd_ids since they don't need to be checked again.
+            self.osd_ids = self.osd_ids.difference(self.emptying_osds)
+
+            # 2) reweight the ok-to-stop osds, ONCE
+            self.reweight_osds(self.emptying_osds)
+
+            # 3) check for osds to be empty
+            empty_osds = self.empty_osds(self.emptying_osds)
+
+            # remove osds that are marked as empty
+            self.emptying_osds = self.emptying_osds.difference(empty_osds)
+
+            # fixed sleep interval of 10 seconds
+            sleep_interval = 10
+            self.log.debug('Sleeping for %d seconds', sleep_interval)
+            self.event.wait(sleep_interval)
+            self.event.clear()
+
+    def shutdown(self):
+        """
+        This method is called by the mgr when the module needs to shut
+        down (i.e., when the serve() function needs to exit).
+        """
+        self.log.info('Stopping')
+        self.run = False
+        self.event.set()
+
+    def osds_not_in_cluster(self, osd_ids: List[int]) -> Set[int]:
+        self.log.info(f"Checking if provided osds <{osd_ids}> exist in the cluster")
+        osd_map = self.get_osdmap()
+        cluster_osds = [x.get('osd') for x in osd_map.dump().get('osds', [])]
+        not_in_cluster = set()
+        for osd_id in osd_ids:
+            if int(osd_id) not in cluster_osds:
+                self.log.error(f"Could not find {osd_id} in cluster")
+                not_in_cluster.add(osd_id)
+        return not_in_cluster
+
+    def empty_osds(self, osd_ids: Set[int]) -> List[int]:
+        if not osd_ids:
+            return list()
+        osd_df_data = self.osd_df()
+        empty_osds = list()
+        for osd_id in osd_ids:
+            if self.is_empty(osd_id, osd_df=osd_df_data):
+                empty_osds.append(osd_id)
+        return empty_osds
+
+    def osd_df(self) -> dict:
+        # TODO: this should be cached I think
```

Comment:
agreed, let's avoid making multiple redundant calls

Reply:
yep, I'd like to postpone this to a later PR as I need to get this merged rather soonish

Reply:
sounds good